sdgym/datasets.py

```diff
@@ -62,94 +95,161 @@ def _get_dataset_path(
     if dataset.exists():
         return dataset
 
-    datasets_path = datasets_path or DATASETS_PATH
+    datasets_path = datasets_path or DATASETS_PATH / modality
     dataset_path = datasets_path / dataset
-    if dataset_path.exists():
+    if dataset_path.exists() and _path_contains_data_and_metadata(dataset_path):
         return dataset_path
 
     bucket = bucket or BUCKET
     if not bucket.startswith(S3_PREFIX):
-        local_path = Path(bucket) / dataset if bucket else Path(dataset)
-        if local_path.exists():
+        local_path = Path(bucket) / modality / dataset if bucket else Path(dataset)
+        if local_path.exists() and _path_contains_data_and_metadata(local_path):
             return local_path
 
-    _download_dataset(
-        modality,
-        dataset,
-        dataset_path,
-        bucket=bucket,
-        aws_access_key_id=aws_access_key_id,
-        aws_secret_access_key=aws_secret_access_key,
+    dataset_path = _download_dataset(
+        modality, dataset, dataset_path, bucket, aws_access_key_id, aws_secret_access_key
     )
     return dataset_path
 
 
-def _get_dataset_subset(data, metadata_dict):
-    if 'tables' in metadata_dict.keys():
-        raise ValueError('limit_dataset_size is not supported for multi-table datasets.')
+def _validate_modality(modality):
+    if modality not in MODALITIES:
+        modalities_list = ', '.join(MODALITIES)
+        raise ValueError(f'Modality `{modality}` not recognized. Must be one of {modalities_list}')
+
+
+def get_data_and_metadata_from_path(dataset_path, modality):
+    """Load dataset data and metadata from a given local path.
 
-    max_rows, max_columns = (1000, 10)
-    columns = metadata_dict['columns']
-    if len(columns) > max_columns:
-        columns = dict(itertools.islice(columns.items(), max_columns))
-        metadata_dict['columns'] = columns
-        data = data[columns.keys()]
+    Args:
+        dataset_path (Path):
+            The local path to the dataset directory containing data and metadata files.
+        modality (str):
+            The dataset modality. Used by the data-reading function to correctly
+            interpret and load the dataset (e.g., ``'single_table'``, ``'multi_table'``).
 
-    data = data.head(max_rows)
+    Returns:
+        tuple[pd.DataFrame | dict, dict | None]:
+            A tuple containing:
+                - The loaded dataset as a ``pandas.DataFrame`` or a dictionary with
+                  ``pandas.DataFrame`` if ``multi_table``.
+            - The dataset metadata as a dictionary.
+    """
+    metadata_dict = None
+    data = None
+    for file_name in dataset_path.iterdir():
+        if 'metadata' in file_name.stem and file_name.suffix == '.json':
+            metadata_dict = _read_metadata_json(dataset_path / file_name)
```

Comment:
Doesn't iterdir return the entire path? So file_name is already the whole path, you shouldn't add dataset_path to it again.

Reply:
Yeah it does :) I just confuse it a lot with `os.listdir` and the old `os.path.join` 😉